### PR TITLE
Closes #364 | Implement dataloader for LTI langID Corpus

### LINKED ISSUE #364: Create dataset loader for LTI LangID Corpus
Dataloader name: `lti_langid_corpus/lti_langid_corpus.py`
DataCatalogue: http://seacrowd.github.io/seacrowd-catalogue/card.html?lti_langid_corpus

|     Dataset| lti_langid_corpus |
|-------------|---|
| Description | The LTI LangID corpus is a dataset for language identification. The most recent version, v5, contains training data for 1266 languages, and some (possibly very tiny) amount of text for a total of 1706 languages. The corpus defines a train:test split, and, when the language in question has a sufficient amount of data, a dev split as well. |
| Subsets | - |
| Languages | ifa, ace, btm, mqj, mbs, nbq, gor, zyp, tbl, kjp, kmk, kqe, ptu, blw, ceb, prf, yva, zlm, bps, tdt, mya, dgc, lus, wrs, abx, rgu, aaz, agn, ccp, cmn, jav, obo, due, msk, xsb, syb, ind, lbk, min, smk, att, nod, tdj, atb, atd, pag, hvn, ksc, lao, kkl, lti, dao, cgc, tbk, gdg, amk, mbd, clu, msb, sbl, cek, khm, yue, sgb, beu, eip, ifu, mnw, suc, bgs, pam, ebk, eng, nfa, cbk, ify, csy, heg, shn, mta, mbt, lex, tha, mmn, sun, vie, llg, xnn, txq, bcl, kje, kne, san, hlt, kyu, bkd, duo, tet, ury, yka, bjn, tiy, ivv, agt, ban, blz, mbi, ilo, mkn, isd, cth, bpr, por, mbb, tgl, msm, ivb, tam, plw, alp, row |
| Tasks | Language Identification, Language Modeling |
| License | Other (other) |
| Homepage | https://www.cs.cmu.edu/~ralf/langid.html |
| HF URL | - |
| Paper URL | https://aclanthology.org/D14-1069.pdf |

### PR DESCRIPTION
**Pull request by ssun32**
Closes #364

### Checkbox
- [ ] Confirm that this PR is linked to the dataset issue.
- [ ] Create the dataloader script `seacrowd/sea_datasets/my_dataset/my_dataset.py` (please use only lowercase and underscore for dataset naming).
- [ ] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_SEACROWD_VERSION` variables.
- [ ] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [ ] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `SEACrowdConfig` for the source schema and one for a seacrowd schema.
- [ ] Confirm dataloader script works with `datasets.load_dataset` function.
- [ ] Confirm that your dataloader script passes the test suite run with `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py`.
- [ ] If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.

### CODE REVIEW

**Review by holylovenia — CHANGES_REQUESTED** [review]
Hi @ssun32, thank you for your contribution. I haven't done an in-depth checking, but I noticed that you have a subset per language. However, we only need a `source` schema to facilitate returning all data and a `seacrowd` schema to facilitate returning all data because the task is language identification.

cc: Please CMIIW, @MJonibek.

**Comment by ssun32** [comment]
> Hi @ssun32, thank you for your contribution. I haven't done an in-depth checking, but I noticed that you have a subset per language. However, we only need a `source` schema to facilitate returning all data and a `seacrowd` schema to facilitate returning all data because the task is language identification.
> 
> cc: Please CMIIW, @MJonibek.

Hello @holylovenia, thank you so much for taking the time to review this pull request. The issue with having a `source` schema and `seacrowd` to load all languages is that the user would have little control over which languages to load. For example, if a user only needs the "Malay" and "Tamil" subsets, he/she must write a filter function over the entire dataset with 100+ languages to pick out the Malay and Tamil instances. It might be better to allow the user to load the Malay subset and Tamil subset independently and concatenate them together. 

Would it better to modify the data such that `source` and `seacrowd` return all languages, while `source_[lang_id]` and `seacrowd_[lang_id]` return the subsets?

**Comment by holylovenia** [comment]
> > Hi @ssun32, thank you for your contribution. I haven't done an in-depth checking, but I noticed that you have a subset per language. However, we only need a `source` schema to facilitate returning all data and a `seacrowd` schema to facilitate returning all data because the task is language identification.
> > cc: Please CMIIW, @MJonibek.
> 
> Hello @holylovenia, thank you so much for taking the time to review this pull request. The issue with having a `source` schema and `seacrowd` to load all languages is that the user would have little control over which languages to load. For example, if a user only needs the "Malay" and "Tamil" subsets, he/she must write a filter function over the entire dataset with 100+ languages to pick out the Malay and Tamil instances. It might be better to allow the user to load the Malay subset and Tamil subset independently and concatenate them together.
> 
> Would it better to modify the dataloader such that `source` and `seacrowd` return all languages, while `source_[lang_id]` and `seacrowd_[lang_id]` return the subsets?

I see what you mean. Do you suppose that splitting a LID dataset based on its language labels (i.e., `{_DATASETNAME}_{lang}_seacrowd_text` and `{_DATASETNAME}_{lang}_source` subsets) has any merit? If yes, then we can leave it be as long as the `{_DATASETNAME}_seacrowd_text` and `{_DATASETNAME}_source` subsets return all of the data.

**Comment by MJonibek** [comment]
Hello @ssun32, I agree with @holylovenia. I understand your point, but to write the filter function is not a big problem. Also, while using 
`load_dataset('seacrowd/sea_datasets/lti_langid_corpus/lti_langid_corpus.py', name='lti_langid_corpus_source')`

 0 examples generating

image

**Comment by holylovenia** [comment]
> Hello @ssun32, I agree with @holylovenia. I understand your point, but to write the filter function is not a big problem. Also, while using `load_dataset('seacrowd/sea_datasets/lti_langid_corpus/lti_langid_corpus.py', name='lti_langid_corpus_source')`

Wait @MJonibek, now I've been enlightened and I quite understand @ssun32's point after thinking for a while... If we have per-language subsets, it would be easier to have different subset compositions.

For example, if I want to have Indonesian-only languages, I can just use `{_DATASETNAME}_ind_seacrowd_text`,  `{_DATASETNAME}_sun_seacrowd_text`, and  `{_DATASETNAME}_jav_seacrowd_text` without including  `{_DATASETNAME}_mya_seacrowd_text` and  `{_DATASETNAME}_ceb_seacrowd_text`.

Now I'm wondering if we should have per-language subsets for all of the existing and the incoming LID dataloaders. 😂 

Do you think my logic makes sense, @MJonibek @ssun32? Or is there any other idea you'd suggest?

cc: I'm tagging @sabilmakbar in this discussion loop since we were discussing the same thing on Discord.

**Comment by ssun32** [comment]
> Hello @ssun32, I agree with @holylovenia. I understand your point, but to write the filter function is not a big problem. Also, while using `load_dataset('seacrowd/sea_datasets/lti_langid_corpus/lti_langid_corpus.py', name='lti_langid_corpus_source')`
> 
> 0 examples generating
> 
> image

The issue with this dataset is that it requires running a complicated 500+ lines shell script to preprocess the data and I took a naive approach to run that script via subprocess.call, hoping that it would support most operating systems other than Windows.  However, I just realized some of the binaries are not that cross-platform compatible, e.g., the subsample binary only works on my Ubuntu server but not on my Mac. 

@holylovenia @MJonibek Should we should convert this to a local dataset and ask the users run the preprocessing script manually instead?

**Comment by holylovenia** [comment]
> > Hello @ssun32, I agree with @holylovenia. I understand your point, but to write the filter function is not a big problem. Also, while using `load_dataset('seacrowd/sea_datasets/lti_langid_corpus/lti_langid_corpus.py', name='lti_langid_corpus_source')`
> > 0 examples generating
> > image
> 
> The issue with this dataset is that it requires running a complicated 500+ lines shell script to preprocess the data and I took a naive approach to run that script via subprocess.call, hoping that it would support most operating systems other than Windows. However, I just realized some of the binaries are not that cross-platform compatible, e.g., the subsample binary only works on my Ubuntu server but not on my Mac.
> 
> @holylovenia @MJonibek Should we should convert this to a local dataset and ask the users run the preprocessing script manually instead?

@ssun32 May I know what kind of preprocessing is needed? Is it to download the data files or indexing or something else?

**Comment by holylovenia** [comment]
> > Hello @ssun32, I agree with @holylovenia. I understand your point, but to write the filter function is not a big problem. Also, while using `load_dataset('seacrowd/sea_datasets/lti_langid_corpus/lti_langid_corpus.py', name='lti_langid_corpus_source')`
> 
> Wait @MJonibek, now I've been enlightened and I quite understand @ssun32's point after thinking for a while... If we have per-language subsets, it would be easier to have different subset compositions.
> 
> For example, if I want to have Indonesian-only languages, I can just use `{_DATASETNAME}_ind_seacrowd_text`, `{_DATASETNAME}_sun_seacrowd_text`, and `{_DATASETNAME}_jav_seacrowd_text` without including `{_DATASETNAME}_mya_seacrowd_text` and `{_DATASETNAME}_ceb_seacrowd_text`.
> 
> Now I'm wondering if we should have per-language subsets for all of the existing and the incoming LID dataloaders. 😂
> 
> Do you think my logic makes sense, @MJonibek @ssun32? Or is there any other idea you'd suggest?
> 
> cc: I'm tagging @sabilmakbar in this discussion loop since we were discussing the same thing on Discord.

Regarding this, I hope you don't mind waiting for a little more time, @ssun32. We're going to have an internal discussion on what subsets LID dataloaders should consist of. I'll get back to you soon.

**Comment by holylovenia** [comment]
> > > Hello @ssun32, I agree with @holylovenia. I understand your point, but to write the filter function is not a big problem. Also, while using `load_dataset('seacrowd/sea_datasets/lti_langid_corpus/lti_langid_corpus.py', name='lti_langid_corpus_source')`
> > 
> > 
> > Wait @MJonibek, now I've been enlightened and I quite understand @ssun32's point after thinking for a while... If we have per-language subsets, it would be easier to have different subset compositions.
> > For example, if I want to have Indonesian-only languages, I can just use `{_DATASETNAME}_ind_seacrowd_text`, `{_DATASETNAME}_sun_seacrowd_text`, and `{_DATASETNAME}_jav_seacrowd_text` without including `{_DATASETNAME}_mya_seacrowd_text` and `{_DATASETNAME}_ceb_seacrowd_text`.
> > Now I'm wondering if we should have per-language subsets for all of the existing and the incoming LID dataloaders. 😂
> > Do you think my logic makes sense, @MJonibek @ssun32? Or is there any other idea you'd suggest?
> > cc: I'm tagging @sabilmakbar in this discussion loop since we were discussing the same thing on Discord.
> 
> Regarding this, I hope you don't mind waiting for a little more time, @ssun32. We're going to have an internal discussion on what subsets LID dataloaders should consist of. I'll get back to you soon.

Hi @ssun32, for LID dataloaders, we agreed upon having an aggregate `source` schema to facilitate returning all data and an aggregate `seacrowd` schema to facilitate returning all data. No need for per-language subsets. If future users need to filter some language labels, we will let them do it after loading the data.

We will check the other LID dataloaders to ensure that the language labels are also in the same format as used here, i.e., [3-letter ISO code](https://docs.google.com/spreadsheets/d/1FX5oNdP2RcPseatKtXIKHfoppi_c986ZrV_gEX-ccgA/edit?usp=sharing).

cc: @MJonibek @SamuelCahyawijaya @sabilmakbar

**Comment by holylovenia** [comment]
Hi @ssun32, please let us know if you need any help for this PR! 👍

**Comment by holylovenia** [comment]
Hi @ssun32, we from SEACrowd are currently looking forward to using your dataloader in our experiments. Do you think it will be possible for you to address the questions and/or requested changes? 😂 We aim to merge it by the end of this week.

File: seacrowd/sea_datasets/lti_langid_corpus/lti_langid_corpus.py [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+
+        lang_id = self.config.subset_id.split("_")[-1]
+
+        datasplits_dir = os.path.join(data_dir, "datasplits")
+
+        if not Path(datasplits_dir).exists() or not Path(os.path.join(datasplits_dir, "DONE")).exists():
+            # Run provided install.sh to generate train/dev/test splits in "datasplits" folder
+            subprocess.call([os.path.join(data_dir, "code", "install.sh"), datasplits_dir])
+            with open(os.path.join(datasplits_dir, "DONE"), "w"):
+                pass
```
[the remark this answers is not in the document]
Comment (ssun32) [reply seacrowd/sea_datasets/lti_langid_corpus/lti_langid_corpus.py:236]: My current implementation is a hack that uses subprocess.call to execute install.sh from the dataset's cache folder but this would not work for people who are not running this from a bash environment. Unfortunately, I no longer have the bandwidth to convert this to something with better cross-platform compatibility. 
I can release this dataset if someone else wants to give it a try. @holylovenia

[the remark this answers is not in the document]
Comment (holylovenia) [reply seacrowd/sea_datasets/lti_langid_corpus/lti_langid_corpus.py:236]: I think it's fine for now. This hack will be helpful for the time being. Thanks for having thought of it. 😁 

Could you please add information regarding "uses subprocess.call to execute install.sh from the dataset's cache folder but this would not work for people who are not running this from a bash environment" in the `_DESCRIPTION`?

File: seacrowd/sea_datasets/lti_langid_corpus/lti_langid_corpus.py [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+
+        train_filepaths = []
+        dev_filepaths = []
+        test_filepaths = []
+
+        for lang_id in _LANGUAGES:
+            train_filepaths.append(
+                (
+                    lang_id,
+                    glob(os.path.join(data_dir, "dataset", "train", f"{lang_id}*")),
```
Comment (holylovenia) [inline seacrowd/sea_datasets/lti_langid_corpus/lti_langid_corpus.py:246]: `glob(os.path.join(datasplits_dir, "train", f"{lang_id}*")),`

File: seacrowd/sea_datasets/lti_langid_corpus/lti_langid_corpus.py [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+                )
+            )
+            dev_filepaths.append(
+                (
+                    lang_id,
+                    glob(os.path.join(data_dir, "dataset", "devtest", f"{lang_id}*")),
```
Comment (holylovenia) [inline seacrowd/sea_datasets/lti_langid_corpus/lti_langid_corpus.py:252]: `glob(os.path.join(datasplits_dir, "devtest", f"{lang_id}*")),`

File: seacrowd/sea_datasets/lti_langid_corpus/lti_langid_corpus.py [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+                )
+            )
+            test_filepaths.append(
+                (
+                    lang_id,
+                    glob(os.path.join(data_dir, "dataset", "test", f"{lang_id}*")),
```
Comment (holylovenia) [inline seacrowd/sea_datasets/lti_langid_corpus/lti_langid_corpus.py:258]: `glob(os.path.join(datasplits_dir, "test", f"{lang_id}*")),`

File: seacrowd/sea_datasets/lti_langid_corpus/lti_langid_corpus.py [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+                    except UnicodeDecodeError:
+                        continue
+
+        elif self.config.schema == "seacrowd_text":
+            key = 0
+            for lang_id, filepaths in all_filepaths:
+                for filepath in filepaths:
+                    try:
+                        for line in open(filepath):
+                            yield key, {"id": key, "text": line.strip(), "label": lang_id}
```
Comment (holylovenia) [inline seacrowd/sea_datasets/lti_langid_corpus/lti_langid_corpus.py:305]: `"id": key,` --> `"id": f"{filepath.split('/')[-1]}_{key}",` for clarity.

**Review by holylovenia — CHANGES_REQUESTED** [review]
Thank you for the fix, @ssun32! Here are my suggestions:

1. I find some samples have `"\u200b"` in place of a space (example below). Could we replace them with spaces instead?
Example:
```
Test 10000
{'id': 'mya_MM.mymr.Burmese.PD-commonlang-test.utf8_10000',
 'label': 20,
 'text': 'ဗိ\u200bမာန်\u200bတော်\u200bတွင်\u200bထာ\u200bဝ\u200bရ\u200bဘု\u200bရား\u200b၏\u200bထံ\u200bတော် '
         'သို့\u200bချဉ်း\u200bကပ်\u200b၍ '
         'အ\u200bမှု\u200bတော်\u200bကို\u200bထမ်း\u200bဆောင်\u200bသော '
         'ယဇ်\u200bပု\u200bရော\u200bဟိတ်\u200bတို့\u200bအ\u200bတွက်\u200bသီး\u200bသန့်\u200bထား\u200bရာ '
         'သန့်\u200bရှင်း\u200bမြင့်\u200bမြတ်\u200bသည့်\u200bဌာ\u200bန\u200bပါ\u200bရှိ\u200b၍ '
         'သူ\u200bတို့ '
         'နေ\u200bထိုင်\u200bရန်\u200bအိမ်\u200bများ\u200bရှိ\u200bရာ\u200bအ\u200bပိုင်း\u200bနှင့်\u200bဗိ\u200bမာန် '
         'တော်\u200bတည်\u200bရှိ\u200bရာ\u200bအ\u200bပိုင်း\u200bပါ\u200bရှိ\u200bလိမ့်\u200bမည်။-'}
```

File: seacrowd/sea_datasets/lti_langid_corpus/lti_langid_corpus.py (hunk 2 above)
Comment (MJonibek) [reply seacrowd/sea_datasets/lti_langid_corpus/lti_langid_corpus.py:246]: @holylovenia I agree to add additional information to `_DESCRIPTION`

**Comment by MJonibek** [comment]
@holylovenia I agree, additional information in `_DESCRIPTION` is good idea

**Comment by holylovenia** [comment]
Hi @ssun32, please let us know if you need any help for this PR! 👍

**Review by holylovenia — APPROVED** [review]
Thank you for the changes, @ssun32! Let's wait for @MJonibek's review.

**Review by MJonibek — APPROVED** [review]
@ssun32 LGTM, thank you for your work
